examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2926 @@
+    Tensor<1, 3>   initial_1d_state;
+  };
+
+  // @sect4{The <code>TimeStep</code> class}
+  //
+  // With the <code>OfflineData</code> and <code>ProblemDescription</code>
+  // classes at hand we can now implement the explicit time-stepping scheme
+  // that was introduced in the discussion above. The main method of the
+  // <code>TimeStep</code> class is <code>step(vector_type &U, double
+  // t)</code> that takes a reference to a state vector <code>U</code> and
+  // a time point <code>t</code> (as input arguments) computes the updated
+  // solution, stores it in the vector <code>temp</code>, swaps its contents
+  // with the vector <code>U</code>, and returns the chosen step-size
+  // $\tau$.
+  //
+  // The other important method is <code>prepare()</code> which primarily
+  // sets the proper partition and sparsity pattern for the temporary
+  // vector <code>temp</code> and the matrix <code>dij_matrix</code>
+  // respectively.
+
+  template <int dim>
+  class TimeStep : public ParameterAcceptor
+  {
+  public:
+    static constexpr unsigned int problem_dimension =
+      ProblemDescription<dim>::problem_dimension;
+
+    using rank1_type = typename ProblemDescription<dim>::rank1_type;
+    using rank2_type = typename ProblemDescription<dim>::rank2_type;
+
+    typedef std::array<LinearAlgebra::distributed::Vector<double>,
+                       problem_dimension>
+      vector_type;
+
+    TimeStep(const MPI_Comm &          mpi_communicator,
+             TimerOutput &             computing_timer,
+             const OfflineData<dim> &  offline_data,
+             const InitialValues<dim> &initial_values,
+             const std::string &       subsection = "TimeStep");
+
+    void prepare();
+
+    double step(vector_type &U, double t);
+
+  private:
+    const MPI_Comm &mpi_communicator;
+    TimerOutput &   computing_timer;
+
+    SmartPointer<const OfflineData<dim>>   offline_data;
+    SmartPointer<const InitialValues<dim>> initial_values;
+
+    SparseMatrix<double> dij_matrix;
+
+    vector_type temp;
+
+    double cfl_update;
+  };
+
+  // @sect4{The <code>SchlierenPostprocessor</code> class}
+  //
+  // At its core, the Schlieren class implements the class member
+  // <code>compute_schlieren</code>. The main purpose of this class member
+  // is to compute an auxiliary finite element field
+  // <code>schlieren</code>, that is defined at each node by
+  // \f[ \text{schlieren}[i] = e^{\beta \frac{ |\nabla r_i|
+  // - \min_j |\nabla r_j| }{\max_j |\nabla r_j| - \min_j |\nabla r_j| } }, \f]
+  // where $r$ can in principle be any scalar quantitiy, in practice
+  // though, the density is a natural candidate, viz. $r := \rho$.
+  // Schlieren postprocessing is a standard method for enhancing the
+  // contrast of a visualization inspired by actual experimental X-ray and
+  // shadowgraphy techniques of visualization.
+
+  template <int dim>
+  class SchlierenPostprocessor : public ParameterAcceptor
```

Comment:
What's the reason why you don't use the `DataPostprocessor` framework for this? It seems unnecessary to first compute a whole vector here, press it into a `DoFHandler` with an associated finite element, and then output that. Or is it the fact that you need the global minimum over $j$ in the computation? That's something you could just compute up front in each time step, and then give it as argument to the `DataPostprocessor` upon creation.

Reply:
Done!

Reply:
See longer discussion below.